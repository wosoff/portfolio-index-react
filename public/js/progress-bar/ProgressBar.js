import React from 'react';
import { BarContainer, Bar } from './style-progress-bar';

const state = {
  count: 0,
};


function changeBarWidth() {
  const bar = document.getElementById('bar');
  const percentText = document.querySelector('span')
  if (bar === null) {
    throw new Error('Invalid ID');
  }

  if (percentText === null) {
    throw new Error('Invalid ID to percent text')
  }

  if (state.count < 100) {
    const duration = 10;
    const timerID = setInterval(() => {
      state.count += 1;

      bar.style.width = `${state.count}%`;
      percentText.textContent = `${state.count}%`;

      if (state.count >= 100) {
        clearInterval(timerID);
        state.count = 0;
      }
    }, duration);
  }
}

export default function Progress() {
  return (
    <>
      <BarContainer>
        <Bar id="bar"/>
        <span>{state.count}%</span>
      </BarContainer>
      <br />
      <button type="button" onClick={changeBarWidth}>Click</button>
    </>
  );
}
