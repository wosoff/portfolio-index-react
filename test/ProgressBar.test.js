import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import renderer from 'react-test-renderer';
import ProgressBar from '../public/js/progress-bar/ProgressBar'

jest.useFakeTimers();

describe('ProgressBar', () => {
  let container = null
  beforeAll(() => {
    container = document.createElement('div')
    document.body.appendChild(container)
  })

  afterAll(() => {
    unmountComponentAtNode(container)
    container.remove()
    container = null;
  })

  test('Snapshot test', () => {
    // For snapshot test by react-test-renderer
    const tree = renderer
      .create(<ProgressBar/>)
      .toJSON();
    
    expect(tree).toMatchSnapshot();
  })

  test(`When it clicks button, width of progress bar changes from 
  0% to 100% for test time;It is 2seconds.`, (done) => {
    act(() => {
      render(<ProgressBar/>, container)
    })

    const button = document.body.querySelector('button')
    const bar = document.body.querySelector('#bar')

    act(() => {
      button.dispatchEvent(new MouseEvent("click", { bubbles: true }));
      
      jest.advanceTimersByTime(2 * 1000);
    })

    expect(bar.style.width).toEqual("100%")
    done()
  })
})