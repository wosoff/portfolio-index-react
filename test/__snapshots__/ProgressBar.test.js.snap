// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar Snapshot test 1`] = `
Array [
  <div
    className="sc-bdVaJa gPWuut"
  >
    <div
      className="sc-bwzfXH kpsefM"
      id="bar"
    />
  </div>,
  <br />,
  <button
    onClick={[Function]}
    type="button"
  >
    Click
  </button>,
]
`;

exports[`ProgressBar When it clicks button, width of progress bar changes from 
  0% to 100%. 1`] = `
Array [
  <div
    className="sc-bdVaJa gPWuut"
  >
    <div
      className="sc-bwzfXH kpsefM"
      id="bar"
    />
  </div>,
  <br />,
  <button
    onClick={[Function]}
    type="button"
  >
    Click
  </button>,
]
`;
